feat(insights): hide "other" breakdown

mariusandra opened the pull request:
## Problem

Not everyone is happy with the "Other" column.

## Changes

Adds a way to remove the column when you edit the chart.

![2023-12-18 14 41 51](https://github.com/PostHog/posthog/assets/53387/b651936b-b03e-469c-9307-e65c2646ad30)


I had to make NULL / None handling for breakdowns very explicit as well. Because of our column materialization, we must also treat `""` as None.

## How did you test this code?

Backend tests

mariusandra committed:
feat(trends): support numeric breakdowns

mariusandra committed:
fix test

mariusandra committed:
feat(insights): hide "other" in breakdowns

mariusandra committed:
Merge branch 'master' into breakdown-hide-other

mariusandra committed:
show in the interface

mariusandra committed:
fix

webjunkie commented:
Reminds me of Kibana:

<img width="527" alt="image" src="https://github.com/PostHog/posthog/assets/59713/2b7c13f9-e9b0-4c36-bb46-2b1678d87af5">

Maybe we'll also do a toggle?

mariusandra committed:
fix schema

mariusandra committed:
fix filter

mariusandra committed:
support nulls in breakdown

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
less changed snapshots!

mariusandra committed:
Merge remote-tracking branch 'origin/breakdown-hide-other' into breakdown-hide-other

mariusandra committed:
swap to a switch

mariusandra commented on posthog/queries/trends/sql.py:
```diff
@@ -98,7 +98,6 @@
 """
 
 TOP_ELEMENTS_ARRAY_OF_KEY_SQL = """
-SELECT groupArray(value) FROM (
```
Running through `groupArray()` removes the rows where this field is `None`. We however do want to know if `None` is part of the returned fields or not.

mariusandra commented on posthog/queries/breakdown_props.py:
```diff
@@ -233,7 +233,12 @@ def get_breakdown_prop_values(
         filter=filter,
         team_id=team.pk,
-    )[0][0]
+    )
+
+    if filter.using_histogram:
+        return response[0][0]
+    else:
+        return [row[0] for row in response]
```
There are two queries here: 1) integer histogram breakdown, 2) string top results breakdown.

The response of the second query changed in order to also return nulls. More on it below.

mariusandra commented on posthog/queries/trends/breakdown.py:
```diff
@@ -83,6 +84,8 @@
 
 BREAKDOWN_OTHER_STRING_LABEL = "$$_posthog_breakdown_other_$$"
 BREAKDOWN_OTHER_NUMERIC_LABEL = 9007199254740991  # pow(2, 53) - 1, for JS compatibility
+BREAKDOWN_NULL_STRING_LABEL = "$$_posthog_breakdown_null_$$"
+BREAKDOWN_NULL_NUMERIC_LABEL = 9007199254740990  # pow(2, 53) - 2, for JS compatibility
```
I'm not a giant fan of magic strings, but for now I consider this to be good enough... Or at least "helps us today and doesn't block us from adding a proper fix in the future".

mariusandra committed:
fix tests

mariusandra committed:
fix tests

github-actions[bot] committed:
Update query snapshots

mariusandra commented:
Finally ready for a review

webjunkie commented on frontend/src/scenes/trends/Trends.tsx:
```diff
@@ -66,27 +66,43 @@ export function TrendInsight({ view, context }: Props): JSX.Element {
+                                                !breakdown.breakdown_hide_other_aggregation,
+                                        })
+                                    }
+                                    label='Group renaming values under "Other"'
```
```suggestion
                                    label='Group remaining values under "Other"'
```

webjunkie reviewed (approved):
Ah, I found it, but only with HogQL trends turned off. Is that as intended?

Otherwise, seems to work alright.

mariusandra committed:
Merge branch 'master' into breakdown-hide-other

mariusandra committed:
label typo

mariusandra committed:
breakdown other

mariusandra commented:
Thanks for the review!

Yes, for now this is only implemented for the non-HogQL trends. I have added "Breakdown controls" as one of the things to fix for the [rewrite](https://github.com/PostHog/meta/issues/130).

I also put an end to the frontend innovation, and moved the button into the sidebar with the other editor filters:

![2023-12-19 09 33 18](https://github.com/PostHog/posthog/assets/53387/df2f317d-2a9a-4781-9a59-4086260e7eb4)

There will likely be some changed snapshots because of this.

One of the next things we can add here is the option to toggle exactly how many breakdown values are displayed, and which ones, but that's for future PRs.

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

thmsobrmlr commented:
Fly-by suggestion: We already have some options relating to breakdowns in the `...`-menu. Should we put the toggle there? I'd also add the breakdown limit into it.

<img width="408" alt="Screenshot 2023-12-19 at 09 43 16" src="https://github.com/PostHog/posthog/assets/1851359/e8cbcff3-43dd-4294-bed1-2d8b1cd19cc6">

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

mariusandra committed:
move to breakdown tag menu

mariusandra commented:
Something like this?
![2023-12-19 10 05 46](https://github.com/PostHog/posthog/assets/53387/b20ef9fd-b3a6-4e1b-a01c-889f9154fa38)

mariusandra committed:
Merge remote-tracking branch 'origin/breakdown-hide-other' into breakdown-hide-other

github-actions[bot] committed:
Update UI snapshots for `webkit` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

mariusandra committed:
cleanup

mariusandra committed:
Merge branch 'master' into breakdown-hide-other

mariusandra committed:
Merge branch 'breakdown-hide-other' of github.com:PostHog/posthog into breakdown-hide-other

thmsobrmlr commented:
> Something like this? ![2023-12-19 10 05 46](https://private-user-images.githubusercontent.com/53387/291534588-b20ef9fd-b3a6-4e1b-a01c-889f9154fa38.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5NzgzNDMsIm5iZiI6MTcwMjk3ODA0MywicGF0aCI6Ii81MzM4Ny8yOTE1MzQ1ODgtYjIwZWY5ZmQtYjNhNi00ZTFiLWEwMWMtODg5ZjkxNTRmYTM4LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDA5MjcyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmNTRmMWRkYTQ4NjM4ODc5YTEwM2MzYjVmOGZjZjdlNzhiNTM5ZDYwZTFlNTliMWUyNzQ1MGIyMzg0ZTJmNzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1ts_NKDXGvHOgBtmMfzsj2UYwQQbgrNlzEd1QX-UwBU) 

Exactly this.

webjunkie commented on frontend/src/scenes/insights/filters/BreakdownFilter/BreakdownTagMenu.tsx:
```diff
@@ -85,6 +91,58 @@ export const BreakdownTagMenu = (): JSX.Element => {
+                                min={1}
+                                value={breakdownLimit}
+                                onChange={(newValue) => {
+                                    setBreakdownLimit(newValue ?? 25)
```
What doesn't work here for me is typing a value and then focusing somewhere else, thereby closing the menu. That doesn't seem to trigger onChange.

mariusandra replied:
Indeed, the UX here is poor. I'll improve it in the next PR as it seems I made the breakdown pills huge now 
<img width="292" alt="image" src="https://github.com/PostHog/posthog/assets/53387/a2b60f69-173b-4598-b954-296e85aab567">
